maup/repair.py

```diff
@@ -1,27 +1,1217 @@
+
+    # Use data about the holes to fill holes if desired.
+    if fill_gaps == True:
+        print("Filling gaps...")
+        reconstructed_df = close_gaps(reconstructed_df, holes_df)
+
+    # Check for precincts that have become (more) disconnected, generally with an extra
+    # component of negligible area.  If any are found and the area is negligible,
+    # reassign to an adjacent precinct by shared perimeter.
+    # If the area is not negligible, leave it alone and report it so that a human
+    # can decide what to do about it!
+
+    disconnected_df = reconstructed_df[
+        reconstructed_df["geometry"].apply(lambda x: x.geom_type != "Polygon")
+    ]
+    # This will include precincts that were disconnected in the original; need to
+    # filter by whether they got worse.
+
+    if len(disconnected_df) > 0:
+        disconnected_poly_indices = []
+        for ind in disconnected_df.index:
+            if num_components(reconstructed_df["geometry"][ind]) > num_components(
+                geometries0_df["geometry"][ind]
+            ):
+                disconnected_poly_indices.append(ind)
+
+        if (
+            len(disconnected_poly_indices) > 0
+        ):  # These are the ones (if any) that got worse.
+            geometries = get_geometries(reconstructed_df)
+            spatial_index = STRtree(geometries)
+            index_by_iloc = dict(
+                (i, list(geometries.index)[i]) for i in range(len(geometries.index))
+            )  # Build indexing dictionary
+
+            for g_ind in disconnected_poly_indices:
+                excess = num_components(
+                    reconstructed_df["geometry"][g_ind]
+                ) - num_components(geometries0_df["geometry"][g_ind])
+
+                component_num_list = [
+                    x for x in range(len(reconstructed_df["geometry"][g_ind].geoms))
+                ]
+                component_areas = []
+
+                for c_ind in range(len(reconstructed_df["geometry"][g_ind].geoms)):
+                    component_areas.append(
+                        (c_ind, reconstructed_df["geometry"][g_ind].geoms[c_ind].area)
+                    )
+
+                component_areas_sorted = sorted(component_areas, key=lambda tup: tup[1])
+
+                big_area = max(
+                    [
+                        reconstructed_df["geometry"][g_ind].area,
+                        geometries0_df["geometry"][g_ind].area,
+                    ]
+                )
+
+                for i in range(
+                    excess
+                ):  # Check that the ith smallest component has small enough area find a better polygon to add it to.
+                    c_ind = component_areas_sorted[i][0]
+                    this_fragment = reconstructed_df["geometry"][g_ind].geoms[c_ind]
+                    if (
+                        component_areas_sorted[i][1] < 0.0001 * big_area
```

Comment:
This threshold seems fairly arbitrary. Should we make it a parameter or at least define it as a top-level constant?